samcli/lib/sync/sync_flow.py

```diff
@@ -111,6 +111,13 @@ def gather_resources(self) -> None:
         """
         raise NotImplementedError("gather_resources")
 
+    def _update_local_hash(self) -> None:
+        if not self._local_sha:
+            LOG.debug("%sNo local hash is configured, skipping to update local hash", self.log_prefix)
+            return
+
+        self._sync_context.update_resource_sync_state(self.sync_state_identifier, self._local_sha)
```

Comment:
Think I missed this in the previous review, but I think the sync context class needs a class-wide thread lock on toml reads and writes to make things atomic, so that we don't skip by accident due to race condition